test/uri/uri_parse_test.cc

```diff
@@ -0,0 +1,117 @@
+#include <gtest/gtest.h>
+#include <sourcemeta/jsontoolkit/uri.h>
+
+// Inspired from
+// https://github.com/uriparser/uriparser/blob/bf0174e83164a4659c51c135399478bec389eafa/test/test.cpp#L302
+
+TEST(URI_parse, syntax_error_1) {
+  EXPECT_THROW(sourcemeta::jsontoolkit::URI uri{"//[::44.1"},
+               sourcemeta::jsontoolkit::URIParseError);
+}
+
+// Inspired from
+// https://github.com/uriparser/uriparser/blob/bf0174e83164a4659c51c135399478bec389eafa/test/test.cpp#L510-L516
+
+TEST(URI_parse, syntax_error_2) {
+  EXPECT_THROW(sourcemeta::jsontoolkit::URI uri{"http://moo:21@moo:21@moo/"},
+               sourcemeta::jsontoolkit::URIParseError);
+}
+
+TEST(URI_parse, syntax_error_3) {
+  EXPECT_THROW(sourcemeta::jsontoolkit::URI uri{"http://moo:21@moo:21@moo:21/"},
+               sourcemeta::jsontoolkit::URIParseError);
+}
+
+// Inspired from
+// https://github.com/uriparser/uriparser/blob/bf0174e83164a4659c51c135399478bec389eafa/test/test.cpp#L2180
+
+TEST(URI_parse, syntax_error_4) {
+  EXPECT_THROW(
+      sourcemeta::jsontoolkit::URI uri{"http://[vA.123456"}, // missing "]"
+      sourcemeta::jsontoolkit::URIParseError);
+}
+
+// Inspired from
+// https://github.com/uriparser/uriparser/blob/bf0174e83164a4659c51c135399478bec389eafa/test/test.cpp#L315
+
+TEST(URI_parse, success_1) {
+  EXPECT_NO_THROW(sourcemeta::jsontoolkit::URI uri{
```

Comment:
Assuming this is parse test, should we assert on the results? At least by recomposing? Otherwise how do we know we are indeed parsing correctly (apart from not throwing?)

Reply:
Yeah I try that :)

I face a few surprises!

Reply:
haha good surprises then